Optimize the chain store block table

jsdanielh opened the pull request:
Optimize the chain store block table by differentiating pushed vs stored blocks in the chain store.
Pushed blocks are partially restored and are rebuilt by using the chain info DB and the history store. Stored blocks are fully stored since there is no associated history.
The `chain_info.on_main_chain` flag is used to differentiate where to store the block and how to rebuild it (if needed).

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
blockchain: Optimize the chain store block table

Optimize the chain store block table by differentiating pushed vs
stored blocks in the chain store.
Pushed blocks are partially restored and are rebuilt by using the
chain info DB and the history store. Stored blocks are fully stored
since there is no associated history.
The `chain_info.on_main_chain` flag is used to differentiate where
to store the block and how to rebuild it (if needed).

sisou commented:
So blocks will be spread across three tables. Do all of them get pruned the same way? So if I request a past block in a full node that pruned that epoch, will I get the correct error? Or half a block?

styppo commented:
These changes would benefit from rebasing them onto #2722. Said PR moves `validators` and `next_batch_initial_punished_set` from the `MacroBody` to the `MacroHeader`, reducing the `MacroBody` to only the reward transactions. Thus, it should be possible to simplify this code by never storing block bodies (including macro blocks) and always recovering them from the `HistoryStore`.

jsdanielh commented:
> So blocks will be spread across three tables.

If you add the history store's history tree table, yes.

> Do all of them get pruned the same way?

No, they get pruned in a different way. For instance, the `stored_block_table` gets pruned on a batch finalization.

> So if I request a past block in a full node that pruned that epoch, will I get the correct error? Or half a block?

The correct error: No block found. The mechanism is the following:
1. Block is looked up in the `store_block_table`. If so, we return it.
2. Block is looked up in the `pushed_block_table`.
  2.1 If so and if it's a micro block, transactions are retrieved from the HS.
3. If both 1 and 2 are a miss an error is returned.

As mentioned before `stored_block_table` is pruned at the end of a batch and `pushed_block_table` might get pruned after an epoch finalization depending on the value of `config.max_epochs_stored` (which is also the case for the HS on full nodes).

jsdanielh commented:
> These changes would benefit from rebasing them onto #2722. Said PR moves `validators` and `next_batch_initial_punished_set` from the `MacroBody` to the `MacroHeader`, reducing the `MacroBody` to only the reward transactions. Thus, it should be possible to simplify this code by never storing block bodies (including macro blocks) and always recovering them from the `HistoryStore`.

Rebased and implemented @styppo

styppo commented on blockchain/src/blockchain/rebranch_utils.rs:
```diff
@@ -248,7 +262,7 @@ impl Blockchain {
             // Include the body of the new block (at position 0).
             self.chain_store
-                .put_chain_info(write_txn, &fork_block.0, &fork_block.1, i == 0);
+                .put_chain_info(write_txn, &fork_block.0, &fork_block.1, i == 0, true);
```
This line looks a bit fishy. We're iterating over `target_chain` here, which are the blocks on the fork we're rebranching to, so these blocks are `stored` in the chain store with full body. When `i == 0` and thus `include_body = true`, the respective block will now also be stored as `pushed`, while all other blocks remain `stored`, even though they should now be considered `pushed`. This doesn't really cause any problems as the blocks can still be discovered, but it is inconsistent and leads to data duplication. However, it might actually be faster to duplicate the transactions instead of deleting them.

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,136 @@
+/// is history associated with this block.
+#[derive(Debug, Serialize, Deserialize, DbSerializable)]
+pub struct PushedBlock {
+    justification: Option<BlockJustification>,
```
The `justification` is already stored as part of `ChainInfo`, so we actually don't need it here.

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,136 @@
+}
+
+impl PushedBlock {
+    fn block_from_header(
+        &self,
+        block: &Block,
```
We could save a bunch of `clone()`s by taking ownership of `self` and `block` here.

```suggestion
    fn into_block(
        self,
        block: Block,
```

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,136 @@
+        txn: &MdbxReadTransaction,
+    ) -> Block {
+        if let Some(justification) = &self.justification {
+            assert!(block.ty() == justification.ty());
```
```suggestion
            assert_eq!(block.ty(), justification.ty());
```

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,136 @@
+            assert!(block.ty() == justification.ty());
+        }
+        if let Some(body) = &self.body {
+            assert!(block.ty() == body.ty());
```
```suggestion
            assert_eq!(block.ty(), body.ty());
```

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,136 @@
+                    .justification
+                    .as_ref()
+                    .map(|justification| justification.clone().unwrap_macro()),
+                body: self.body.as_ref().map(|_| {
```
This is a bit strange. We could just completely ignore `self.body` in this case (or even better, get rid of it entirely for macro blocks)?

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,136 @@
+/// optimize blocks storage. This assumes that a block has been pushed and that there
+/// is history associated with this block.
+#[derive(Debug, Serialize, Deserialize, DbSerializable)]
+pub struct PushedBlock {
+    justification: Option<BlockJustification>,
+    #[serde(serialize_with = "PushedBlock::serialize_body")]
+    body: Option<BlockBody>,
+}
```
```suggestion
pub enum PushedBlock {
    Macro,
    Micro {
        equivocation_proofs: Vec<EquivocationProof>,
    }
}
```
This could be an alternative way to model this. Technically, we wouldn't need to store anything for macro blocks at all when they are `pushed` (which they should always be).

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -95,8 +205,12 @@ impl ChainStore {
         if include_body {
-            if let Some(block) = txn.get(&self.block_table, hash) {
-                chain_info.head = block;
+            // Check the stored block table, then the pushed block table
```
We could eliminate one database fetch operation here by storing a `pushed` flag in `ChainInfo`, which tells us which block table to use.

jsdanielh replied:
Just implemented this but used the already present `chain_info.on_main_chain` flag.

jsdanielh commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -378,7 +378,7 @@ impl Blockchain {
 
         this.chain_store
-            .put_chain_info(&mut txn, &block_hash, &chain_info, true);
+            .put_chain_info(&mut txn, &block_hash, &chain_info, false);
```
@styppo I believe this can be `false` since we are already storing the `chain_info` with `chain_info.on_main_chain` and we're setting `include_body=true` at [line 172](https://github.com/nimiq/core-rs-albatross/pull/2823/files#diff-347ff6e4fe84ae58c6cf527d349b6acc2f21444c88e11efc6deb7028915daec4R172).

jsdanielh committed:
pow-migration: Do not use `Policy` functions to calculate epoch numbers

Change the blockchain history store interface to add a function
(`add_to_history_for_epoch`) that receives the epoch number instead
of relying on `Policy` to calculate it. This is useful in the history
migration for `PoW` since the genesis block number wouldn't be known
until the migration is over.

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,121 @@
+    fn ty(&self) -> BlockType {
+        match self {
+            PushedBlock::Macro => BlockType::Macro,
+            PushedBlock::Micro {
+                equivocation_proofs: _,
+            } => BlockType::Micro,
```
```suggestion
            PushedBlock::Micro { .. } => BlockType::Micro,
```

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -1,21 +1,121 @@
+            } => BlockType::Micro,
+        }
+    }
+    fn into_block(
+        self,
+        block: &Block,
+        history_store: &Arc<HistoryStoreProxy>,
+        txn: &MdbxReadTransaction,
+    ) -> Block {
+        assert_eq!(self.ty(), block.ty());
+        let mut block_to_return = block.clone();
```
```suggestion
    fn populate_body(
        self,
        block: &mut Block,
        history_store: &Arc<HistoryStoreProxy>,
        txn: &MdbxReadTransaction,
    ) {
        assert_eq!(self.ty(), block.ty());
```
We can eliminate the `block.clone()` here by passing a mutable reference to the partial block.

styppo commented on blockchain/src/chain_store.rs:
```diff
@@ -161,7 +287,18 @@ impl ChainStore {
+                txn.put_reserve(
+                    &self.pushed_block_table,
+                    hash,
+                    &PushedBlock::from(chain_info.head.clone()),
```
The `clone` here is unnecessarily expensive since it clones the full block, but `PushedBlock::from` only needs the `equivocation_proofs` for micro blocks. It would be more efficient to implement `From<&Block>` and clone only the `equivocation_proofs` within the trait implementation.

styppo commented:
LGTM
